### ISSUE TITLE
Uploaded file is removed even it is well uploaded (Meteor 1.4.2.3, Meteor-Files 1.7.6)

### ISSUE THREAD

**Issue by tophsic** [VeliovGroup/Meteor-Files#324]
Hi,

As you advise, I searched a solution by myself before submitting this issue :wink: I took quite some time to find a workaround and try to give you maximum information to understand the situation.

I used Meteor-Files 1.6.6 without any problem and I upgraded Meteor from 1.3.4.4 to 1.4.2.3. With this upgrade, I updated Meteor-Files into 1.7.6. I did not noticed anything in development environment but when I deployed in production, each file I uploaded were removed immediately.

I run my application in production with Nginx and Phusion Passenger on Ubuntu 16.04.

```bash
App 26871 stdout: [FilesCollection] [File Start Method] TestFile.xlsx - HoWhviCGcmJ67bSjG
App 26871 stdout: [FilesCollection] [Upload] [Start Method] Got #-1/1 chunks, dst: TestFile.xlsx
App 26871 stdout: [FilesCollection] [Upload] [DDP] Got #1/1 chunks, dst: TestFile.xlsx
App 26871 stdout: [FilesCollection] [Upload] [DDP] Got #-1/1 chunks, dst: TestFile.xlsx
App 26871 stdout: [FilesCollection] [Upload] [finish(ing)Upload] -> assets/path/files/HoWhviCGcmJ67bSjG.xlsx
App 26871 stdout: [FilesCollection] [Upload] [finish(ed)Upload] -> assets/path/files/HoWhviCGcmJ67bSjG.xlsx
App 26871 stdout: [FilesCollection] [_preCollectionCursor.observe] [removed]: HoWhviCGcmJ67bSjG
App 26871 stdout: [FilesCollection] [_preCollectionCursor.observe] [removeUnfinishedUpload]: assets/path/files/HoWhviCGcmJ67bSjG.xlsx
``` 

I downgraded to Meteor-Files 1.7.5 and did not observe this behavior

```bash
App 30506 stdout: [FilesCollection] [File Start Method] TestFile.xlsx - ctt4jBvctNgxRj954
App 30506 stdout: [FilesCollection] [Upload] [Start Method] Got #-1/1 chunks, dst: TestFile.xlsx
App 30506 stdout: [FilesCollection] [Upload] [DDP] Got #1/1 chunks, dst: TestFile.xlsx
App 30506 stdout: [FilesCollection] [Upload] [DDP] Got #-1/1 chunks, dst: TestFile.xlsx
App 30506 stdout: [FilesCollection] [Upload] [finish(ing)Upload] -> assets/path/files/ctt4jBvctNgxRj954.xlsx
App 30506 stdout: [FilesCollection] [Upload] [finish(ed)Upload] -> assets/path/files/ctt4jBvctNgxRj954.xlsx
App 30506 stdout: [FilesCollection] [_preCollectionCursor.observeChanges] [removed]: ctt4jBvctNgxRj954
```

As I debugged on core mode directly in compiled javascript code. I added just following line to see if isFinished flag was well updated. Strangely, that solved the problem.

```javascript
  FilesCollection.prototype._finishUpload = Meteor.isServer ? function (result, opts, cb) {                           //
    var self;                                                                                                         //
    if (this.debug) {                                                                                                 //
      console.info("[FilesCollection] [Upload] [finish(ing)Upload] -> " + result.path);                               //
    }                                                                                                                 //
    fs.chmod(result.path, this.permissions, NOOP);                                                                    //
    self = this;                                                                                                      //
    result.type = this._getMimeType(opts.file);                                                                       //
    result["public"] = this["public"];                                                                                //
    this.collection.insert(_.clone(result), function (error, _id) {                                                   //
      if (error) {                                                                                                    //
        cb && cb(error);                                                                                              //
        if (self.debug) {                                                                                             //
          console.error('[FilesCollection] [Upload] [_finishUpload] Error:', error);                                  //
        }                                                                                                             //
      } else {                                                                                                        //
        self._preCollection.update({                                                                                  //
          _id: opts.fileId                                                                                            //
        }, {                                                                                                          //
          $set: {                                                                                                     //
            isFinished: true                                                                                          //
          }                                                                                                           //
        }, function () {                                                                                              //
          // LINE ADDED TO DEBUG ////////////////////////////////////////////
          console.info(self._preCollection.findOne({ _id: opts.fileId })); //
          ///////////////////////////////////////////////////////////////////
          self._preCollection.remove({                                                                                //
            _id: opts.fileId                                                                                          //
          }, function (error) {                                                                                       //
            if (error) {                                                                                              //
              cb && cb(error);                                                                                        //
              if (self.debug) {                                                                                       //
                console.error('[FilesCollection] [Upload] [_finishUpload] Error:', error);                            //
              }                                                                                                       //
            } else {                                                                                                  //
              result._id = _id;                                                                                       //
              if (self.debug) {                                                                                       //
                console.info("[FilesCollection] [Upload] [finish(ed)Upload] -> " + result.path);                      //
              }                                                                                                       //
              self.onAfterUpload && self.onAfterUpload.call(self, result);                                            //
              self.emit('afterUpload', result);                                                                       //
              cb && cb(null, result);                                                                                 //
            }                                                                                                         //
          });                                                                                                         //
        });                                                                                                           //
      }                                                                                                               //
    });                                                                                                               //
  } : void 0;                                                                                                         //
```

**Comment by tophsic** [comment]
Did not search enough :disappointed:. Can it be possible it is related to a4fb1a6? See #267 

I do not understand why my uploaded file is not considered as finished.

**Comment by dr-dimitru** [comment]
Hello @tophsic ,

Yes, some of developers faced this issue at docker.
You can safely downgrade to `v1.7.5` this issue will be solved.

Although this was reported my multiple developers - I still can't reproduce this issue. Could you please describe your environment and how you deploy app, please?

**Comment by tophsic** [comment]
I deploy my app on dedicated servers:

* web app on ubuntu 16.04 (2 CPU, 4 GO), nginx 1.10.2, phusion passenger 5.1.0
* database on ubuntu 16.04 (1 CPU, 1 GO), mongo 3.2.11

Meteor release: `METEOR@1.4.2.3`

Packages version:

```
accounts-base@1.2.14
accounts-password@1.3.3
aldeed:autoform@5.8.1
aldeed:simple-schema@1.5.3
aldeed:template-extension@3.4.3
allow-deny@1.0.5
aramk:google-maps-marker-label@1.1.10
autoupdate@1.3.12
babel-compiler@6.13.0
babel-runtime@1.0.1
base64@1.0.10
binary-heap@1.0.10
blaze@2.2.0
blaze-html-templates@1.0.5
blaze-tools@1.0.10
boilerplate-generator@1.0.11
caching-compiler@1.1.9
caching-html-compiler@1.0.7
callback-hook@1.0.10
chart:chart@1.0.1-beta.4
check@1.2.4
coffeescript@1.11.1_4
dburles:google-maps@1.1.5
ddp@1.2.5
ddp-client@1.3.2
ddp-common@1.2.8
ddp-rate-limiter@1.0.6
ddp-server@1.3.12
deps@1.0.12
diff-sequence@1.0.7
dispatch:mocha-phantomjs@0.1.9
dispatch:phantomjs-tests@0.0.7
ecmascript@0.6.1
ecmascript-runtime@0.3.15
ejson@1.0.13
email@1.1.18
es5-shim@4.6.15
fastclick@1.0.13
force-ssl@1.0.13
fortawesome:fontawesome@4.7.0
fourseven:scss@3.13.0
geojson-utils@1.0.10
handlebars@1.0.7
hot-code-push@1.0.4
html-tools@1.0.11
htmljs@1.0.11
http@1.2.10
id-map@1.0.9
jplack:bootstrap-timepicker@0.2.2
jquery@1.11.10
jspdf:core@1.1.135
kadira:flow-router@2.12.1
launch-screen@1.1.0
livedata@1.0.18
localstorage@1.0.12
logging@1.1.16
mdg:geolocation@1.3.0
mdg:validated-method@1.1.0
mdg:validation-error@0.5.1
meteor@1.6.0
meteor-base@1.0.4
meteorhacks:kadira@2.30.3
meteorhacks:meteorx@1.4.1
minifier-css@1.2.15
minifier-js@1.2.15
minimongo@1.0.19
mobile-experience@1.0.4
mobile-status-bar@1.0.13
modules@0.7.7
modules-runtime@0.7.7
momentjs:moment@2.17.1
mongo@1.1.14
mongo-id@1.0.6
mongo-livedata@1.0.12
mrt:googlemaps@0.1.1
mrt:reactive-extra@0.1.0
npm-bcrypt@0.9.2
npm-mongo@2.2.11_2
observe-sequence@1.0.14
ordered-dict@1.0.9
ostrio:cookies@2.0.5
ostrio:files@1.7.5
peppelg:bootstrap-3-modal@1.0.4
percolate:migrations@0.9.8
practicalmeteor:chai@2.1.0_1
practicalmeteor:loglevel@1.2.0_2
practicalmeteor:mocha@2.4.5_6
practicalmeteor:mocha-core@1.0.1
practicalmeteor:sinon@1.14.1_2
promise@0.8.8
rajit:bootstrap3-datepicker@1.6.4
rajit:bootstrap3-datepicker-fr@1.6.4
random@1.0.10
rate-limit@1.0.6
reactive-dict@1.1.8
reactive-var@1.0.11
reload@1.1.11
retry@1.0.9
risul:bootstrap-colorpicker@2.3.6
routepolicy@1.0.12
rzymek:moment-locale-fr@2.14.1
seba:minifiers-autoprefixer@1.0.1
service-configuration@1.0.11
session@1.1.7
sha@1.0.9
shell-server@0.2.1
simple:json-routes@2.1.0
spacebars@1.0.13
spacebars-compiler@1.0.13
srp@1.0.10
standard-minifier-js@1.2.1
templating@1.2.15
templating-compiler@1.2.15
templating-runtime@1.2.15
templating-tools@1.0.5
tmeasday:test-reporter-helpers@0.2.1
tracker@1.1.1
twbs:bootstrap@3.3.6
ui@1.0.12
underscore@1.0.10
url@1.0.11
webapp@1.3.12
webapp-hashing@1.0.9
```

Here are deployment steps (I do not use mup yet):

* git clone / update local repository
* meteor build into web
* npm install into `program/server` directory
* restart nginx

**Comment by dr-dimitru** [comment]
@tophsic thank you very much for info.

I see you're using Phusion. It has two possibilities for Meteor apps, run it as "Meteor" and run it as regular Node.js app (when meteor is "compiled" to node.js). Which option you're on?

**Comment by Sphinxxxx** [comment]
Same issue here. `Meteor 1.4.2.3` deployed to a Debian server using https://github.com/kadirahq/meteor-up

Downgrading to 1.7.5 solved it for me too.

**Comment by macrozone** [comment]
Same issue here. Lost a lot of time on this.
Issue does not happen on local development instance, so it might be a nasty timing issue

Here some logs:

```

[FilesCollection] [File Start Method] Picture5.png - ctzkzEKJ8EDDvNndm
[FilesCollection] [Upload] [Start Method] Got #-1/1 chunks, dst: Picture5.png
[FilesCollection] [findOne("ctzkzEKJ8EDDvNndm", undefined)]
[FilesCollection] [Upload] [DDP] Got #1/1 chunks, dst: Picture5.png
[FilesCollection] [findOne("ctzkzEKJ8EDDvNndm", undefined)]
[FilesCollection] [Upload] [DDP] Got #-1/1 chunks, dst: Picture5.png
[FilesCollection] [findOne("ctzkzEKJ8EDDvNndm", undefined)]
[FilesCollection] [Upload] [finish(ing)Upload] -> /home/app/app/shared/ctzkzEKJ8EDDvNndm.png
[FilesCollection] [Upload] [finish(ed)Upload] -> /home/app/app/shared/ctzkzEKJ8EDDvNndm.png
[FilesCollection] [_preCollectionCursor.observe] [removed]: ctzkzEKJ8EDDvNndm
[FilesCollection] [_preCollectionCursor.observe] [removeUnfinishedUpload]: /home/app/app/shared/ctzkzEKJ8EDDvNndm.png


```

**Comment by bobsov** [comment]
Is downgrading to 1.7.5 is the only option here? I am facing the same issue.

**Comment by bedorn** [comment]
Same issue as macrozone and downgrading does not solve it. Finally I am at the end with my latin. I guess that the issue occurs due to docker, as it works properly on my local machine. Would be very helpfuly to have a solution

**Comment by bobsov** [comment]
For me downgrading to 1.7.5 resolves the issue as I am using custom storagePath.

In local instance it is sometimes hit or miss if I don't define custom storagePath and when default upload location is used. Sometimes, I see uploaded file there and sometimes it gets removed. As someone else said in the forum, it is all about timing.

I think whats happening is - the 1.7.6 was meant to remove temp file but it actually ends up removing the actual uploaded file depending on the variable being set.

**Comment by bedorn** [comment]
I am sorry, you are right. Downgrade to 1.7.5 solved it

**Comment by dr-dimitru** [comment]
@bedorn Yes, downgrading to v1.7.5 should solve it.

We are still investigating it. Next release will solve it.

**Comment by derwok** [comment]
On three boxes we had no problems with uploading files.

* One was MacOS with MongoDB 3.2.6
* One was CentOS with MongoDB 3.2.6
* One was Ubuntu with MongoDB 3.2.6

Then came a staging server with a quite outdated MongoDB 2.6.12 - there we could not upload a single file. No matter what we tried. Always this issue #324 stopped the show as it deleted the file after successful 100% upload.

I provided a pull request that delays the collection.remove a little bit. Actually 1 millisecond was enough - I wnt with 25ms... That solved our problem on our server.

Please take a close look that I don't break anything, as I'm not familiar with coffescript.

Thanks for taking care.

**Comment by dr-dimitru** [comment]
Hi @derwok ,

Thank you for this tests and PR.
Left a comment on PR

**Comment by dr-dimitru** [comment]
Hi @tophsic @Sphinxxxx @macrozone @bedorn @bobsov ,

This issue should be fixed in `v1.7.7` (currently on [`dev`](https://github.com/VeliovGroup/Meteor-Files/tree/dev) brunch). I'll be glad if you can tests it on your end before release.

To use package from dev brunch:
```shell
# cd to your project root
mkdir packages
cd packages
git clone https://github.com/VeliovGroup/Meteor-Files.git
cd Meteor-Files
git checkout dev
cd ../../
meteor update # use --all-packages flag if necessary
```

Do remove dev package:
```shell
# cd to your project root
rm -R ./packages/Meteor-Files
```
